### PR TITLE
Add new peaklet classificaiton dtype

### PR DESCRIPTION
**Pull request by ahiguera-mx**
_Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible_
## What does the code in this PR do / what does it improve?
This is a very rough draft 
Add new peaklet classification method based on a Bayes classifier 

## Can you briefly describe how it works?
The idea is to add 
```
+ [('type_bayse', np.int8, 'Bayes peak classification type')]
             + [('s1_prob', np.float32, 'S1 probability' )]
             + [('s2_prob', np.float32, 'S2 probability' )]
```
Then at `event_processing` user can add if it builds events based on p['type'] or p[''type_bayse'] or p['s1/s2_prob']
TO DO:

- modify `peak` class
- modify `peak_basics` class

## Can you give a minimal working example (or illustrate with a figure)?
`st.make(run_ID, 'peaklet_classification')`
_Please include the following if applicable:_
  - [ ] _Update the docstring(s)_
  - [ ] _Update the documentation_
  - [ ] _Tests to check the (new) code is working as desired._
  - [ ] _Does it solve one of the open issues on github?_

### _Notes on testing_
 - _Until the automated tests pass, please mark the PR as a draft._
 - _On the XENONnT fork we test with database access, on private forks there is no database access for security considerations._

All _italic_ comments can be removed from this template.

### CODE REVIEW

**Comment by WenzDaniel** [comment]
Hej, Aaron can we discuss a bit the purpose of the second classification plugin a bit? In the sense that we discuss how we want to compare it to our current classification approach? I think to fully test it within the entire processing framework we also have to include event-building later on etc. 

Hence I think your plugin should do the following:

- Provide the same data_type as the current classification plugin: `peaklet_classification`
- Replace `type` with the `type`-classification of your approach (you can add additional fields, but you have to overwrite `type` so we can check event-building)
- Remove all the code which belongs to the old classification from your plugin. Having this duplicated code is dangerous as changes may not be forwarded and the comparison can become faulty. 
- When comparing the two methods you can use two contexts one using the default one and one context using the Rice approach. 

Please let me know if you agree. We can also have a short meeting this week if you like.

**Comment by ahiguera-mx** [comment]
Hi Daniel,

Thanks for looking into this very quickly! This still a rough draft and mostly to start the discussion, here is what I have in mind so hopefully this clarify things 
The idea that I have in mind is that `peaklet_classification` class return two types, one is going to be the current method and another called type_bayes, propagate this new dtypes to `peaks`, then at the event processing the user can decide whether wants to use the current type method or the new, I thought about this since doing event_processing is very quick to process.
In that sense, there is not duplication I just put the   `peaklet_classification` class on its own. And yes probably having a meeting to discuss this would be better

**Comment by WenzDaniel** [comment]
But this is not what we want right? An analyst should not decide which algorithm to use. There should be just a single classification everyone is using. So we can save us here a lot of trouble and work by just having two plugins as long as the ML validation is going on. After that we register the ML-classification plugin by default.

**Comment by ahiguera-mx** [comment]
The way that I see it is that we are adding more information. In my opinion having this flexibility is better, at the end when we define a given context we can set which peak classification method we want.

**Comment by WenzDaniel** [comment]
But why would you keep a classification which is worse? I do not understand.

**Comment by JoranAngevaare** [comment]
@ahiguera-mx thanks for the PR. I'm sorry but what you describe is best achieved by what Daniel is proposing.

What you are doing here will require a new reprocessing without the chance to first validate and compare.

File: straxen/plugins/peaklet_classification.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    def __init__(self, num_nodes=50, bins=None, CPT=None ):
+        self.num_nodes = num_nodes
+        self.classes = 2
+        self.class_prior = np.array([1. /self.classes for j in range(self.classes)])
+        self.bins = np.load(bins)
+        self.CPT = np.load(CPT)
+
+
+    def compute(self, peaklets):
+        bayse_ptype = np.zeros(len(peaklets), dtype=np.int8)
```
Comment (sophiafarrell) [inline straxen/plugins/peaklet_classification.py:147]: Can we correct this typo?

File: straxen/plugins/peaklet_processing.py [hunk 2]
```diff
@@ -680,8 +588,17 @@ def setup(self):
                                                      self.config['gain_model'])
 
     def infer_dtype(self):
-        return strax.unpack_dtype(self.deps['peaklets'].dtype_for('peaklets'))
-
+        # wrong order
+        dtype = (strax.unpack_dtype(self.deps['peaklets'].dtype_for('peaklets'))
+                 +[('type_bayse', np.int8, 'Bayes peak classification type')]
```
Comment (sophiafarrell) [inline straxen/plugins/peaklet_processing.py:593]: Also can we correct this typo? Causing mailbox errors.

File: straxen/plugins/peaklet_classification.py (hunk 1 above)
Comment (ahiguera-mx) [reply straxen/plugins/peaklet_classification.py:147]: actually we don't need this

File: straxen/plugins/peaklet_processing.py (hunk 2 above)
Comment (ahiguera-mx) [reply straxen/plugins/peaklet_processing.py]: good catch, thanks should be fixed now

File: straxen/plugins/peaklet_processing.py [hunk 3]
```diff
@@ -680,8 +588,17 @@ def setup(self):
                                                      self.config['gain_model'])
 
     def infer_dtype(self):
-        return strax.unpack_dtype(self.deps['peaklets'].dtype_for('peaklets'))
-
+        # wrong order
+        dtype = (strax.unpack_dtype(self.deps['peaklets'].dtype_for('peaklets'))
+                 +[('type_bayes', np.int8, 'Bayes peak classification type')]
+                 +[('s1_prob', np.float32, 'S1 probability' )]
```
Comment (sophiafarrell) [inline straxen/plugins/peaklet_processing.py:594]: need to add ln to doc, otherwise it throws mailbox error (same with line below)

**Comment by sophiafarrell** [comment]
Hey Daniel, Joran, thanks for your fast feedback. 
* Regarding `'type'`, I think we do agree with your suggestion: we need to propagate this classifier through -- otherwise, it dies at the peaklet level.  For now, I've made some edits to Aaron's code such that this processes data for us, to do the proper checks. However, I would like to restructure things as you both suggest.
* Given that above point, and the timing of this PR, I hope it is okay for us to keep this PR draft up so we can continue to tune/implement this. There are some hacks in here to get us to test/debug on data, but we plan to make these more legitimate once we implement your type suggestion.

**Comment by WenzDaniel** [comment]
Hej Sophia, sure no problem. If you have any questions or need some advice just let us know.

**Comment by ahiguera-mx** [comment]
Hi @JoranAngevaare @WenzDaniel 
Do you guys have an example on how to overwrite a given dtype in straxen. For instance, my new plug-in see below
how can I overwrite `type` so I can check at the event level. My first guess is to use `strax.MergeOnlyPlugin` however that would only work if the `dtypes` are identical, notice that in my plug-in I want to add two extra dtypes `s1_prob` and `s2_prob`

```
provides = 'bayes_peaklet_classification' 
depends_on = ('peaklets',) 
parallel = True 
 __version__ = '0.0.1' 
dtype = (strax.peak_interval_dtype 
                 + [('type', np.int8, 'Bayes peak classification type')] 
                 + [('s1_prob', np.float32, 'S1 probability' )] 
                 + [('s2_prob', np.float32, 'S2 probability' )] 
                ) 
```

**Comment by sophiafarrell** [comment]
Hey @JoranAngevaare @WenzDaniel! Re-tagging about the above question Aaron had. 

**More info:**
Our current idea is to just create this bayes classification plugin that calculates the Bayes classisfication and peak class probabilities. We'd then need to override the `type` of peaks, if some argument `use_bayes_classifier=True` (we'd default this to False, for now, until I can do more data-driven checks.) 
I'll link Aaron's preliminary plugin [here](https://github.com/XENONnT/straxen/compare/master...RiceAstroparticleLab:master), we greatly appreciate any review or feedback you can give in addition to info on overwriting `type`, as Aaron asked. Thx in advance!

**Comment by WenzDaniel** [comment]
I am afraid, I am a bit lost here, sorry. Why do you need to overwrite the peak-data_type? You make a new plugin which also provides peaklet-classification with some additional fields. Lets say its name is `PeakltClassifcation2`. Is not it sufficient to only update event_basics accordingly? If you use your context and register `PeakletClassification` the additional fields in event_basics are just zeros. If you use `PeakletClassification2` it is filled with the corresponding values.  You can also add an additional event-level plugin which loads the corresponding information. Might be even a bit cleaner and easier to handle while comparing old and new stuff.

**Comment by tunnell** [comment]
This is public so I encourage just a call just to be effective with everybody's time.  There were specific concerns raised from others that this aims to address, where we can discuss details on Slack.

**Comment by WenzDaniel** [comment]
Sure I am usually available.

**Comment by ahiguera-mx** [comment]
Thanks for all the useful conversation, we are closing this as a new and complete PR is coming.

**Pull request closed without merging**